### PR TITLE
feat: add cache for compute MAC

### PR DESCRIPTION
**Pull request by fqutishat**
Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

### CODE REVIEW

File: pkg/kms/webkms/headers_opts.go [hunk 1]
```diff
@@ -6,30 +6,42 @@ SPDX-License-Identifier: Apache-2.0
...
-func WithHeaders(addHeadersFunc addHeaders) HeadersOpt {
-	return func(opts *HeadersOpts) {
+func WithHeaders(addHeadersFunc addHeaders) Opt {
+	return func(opts *Opts) {
 		opts.HeadersFunc = addHeadersFunc
 	}
 }
+
+// WithCache add cache. if size is zero cache content will not be purged.
+func WithCache(cacheSize int) Opt {
```
Comment (troyronda) [inline pkg/kms/webkms/headers_opts.go:43]: As a followup, the structural issues between encrypted formatter and rest provider should still be resolved. The structure shouldn't depend on the existence of a cache. @fqutishat @DRK3

File: pkg/kms/webkms/headers_opts.go [hunk 2]
```diff
@@ -6,30 +6,42 @@ SPDX-License-Identifier: Apache-2.0
...
 // authorization server to fetch the necessary info needed in these headers).
-func WithHeaders(addHeadersFunc addHeaders) HeadersOpt {
-	return func(opts *HeadersOpts) {
+func WithHeaders(addHeadersFunc addHeaders) Opt {
+	return func(opts *Opts) {
 		opts.HeadersFunc = addHeadersFunc
 	}
 }
+
+// WithCache add cache. if size is zero cache content will not be purged.
```
[the remark this answers is not in the document]
Comment (troyronda) [reply pkg/kms/webkms/headers_opts.go:42]: I would also like to understand better what calls or headers are being duplicated that resulted in a cache, beyond the structure issues. @DRK3 @fqutishat

**Review by troyronda — APPROVED** [review]

File: pkg/crypto/webkms/remotecrypto.go [hunk 3]
```diff
@@ -335,7 +338,18 @@ func (r *RemoteCrypto) Verify(signature, msg []byte, keyURL interface{}) error {
 
 // ComputeMAC remotely computes message authentication code (MAC) for code data with key at keyURL.
 // using a matching MAC primitive in kh key handle.
-func (r *RemoteCrypto) ComputeMAC(data []byte, keyURL interface{}) ([]byte, error) {
+func (r *RemoteCrypto) ComputeMAC(data []byte, keyURL interface{}) ([]byte, error) { //nolint: gocyclo
+	if r.opts.ComputeMACCache != nil {
+		v, err := r.opts.ComputeMACCache.Get(string(data))
```
Comment (troyronda) [inline pkg/crypto/webkms/remotecrypto.go:343]: The cache key should be both data and keyuRL

File: pkg/crypto/webkms/remotecrypto.go [hunk 4]
```diff
@@ -373,6 +387,12 @@ func (r *RemoteCrypto) ComputeMAC(data []byte, keyURL interface{}) ([]byte, erro
 		return nil, err
 	}
 
+	if r.opts.ComputeMACCache != nil {
+		if err := r.opts.ComputeMACCache.Set(string(data), macBytes); err != nil {
```
Comment (troyronda) [inline pkg/crypto/webkms/remotecrypto.go:391]: The cache key should be both data and keyURL.

You might want to consider hashing.

**Review by DRK3 — APPROVED** [review]

File: pkg/crypto/webkms/remotecrypto.go (hunk 3 above)
Comment (fqutishat) [reply pkg/crypto/webkms/remotecrypto.go]: done

File: pkg/crypto/webkms/remotecrypto.go (hunk 4 above)
Comment (fqutishat) [reply pkg/crypto/webkms/remotecrypto.go]: done

**Review by troyronda — APPROVED** [review]

**Pull request merged**